### PR TITLE
ansible,jenkins: install and use GCC 10 on ubuntu2004_sharedlibs

### PR DESCRIPTION
**Pull request by targos**
Refs: https://github.com/nodejs/build/issues/3317

### CODE REVIEW

File: jenkins/scripts/select-compiler.sh [hunk 1]
```diff
@@ -239,6 +239,18 @@ elif [ "$SELECT_ARCH" = "ARM64" ]; then
       fi
       echo "Compiler set to GCC" `$CXX -dumpversion`
       ;;
+    *ubuntu2004_sharedlibs* )
```
Comment (richardlau) [inline jenkins/scripts/select-compiler.sh:242]: We could possibly make this 
```suggestion
    *ubuntu2004* )
```
but I'm okay with it as-is and update if necessary later on.

**Review by richardlau — APPROVED** [review]

**Comment by targos** [comment]
I'm getting an unrelated error while trying to apply it:

```
ansible-playbook --limit test-osuosl-ubuntu2004_docker-arm64-1 ansible/playbooks/jenkins/docker-host.yaml
```

```
TASK [github : write github.com entry in known_hosts] ************************************************************************************************
fatal: [test-osuosl-ubuntu2004_docker-arm64-1]: FAILED! => {"msg": "Failed to set permissions on the temporary files Ansible needs to create when becoming an unprivileged user (rc: 1, err: chmod: invalid mode: ‘A+user:iojs:rx:allow’\nTry 'chmod --help' for more information.\n}). For information on working around this, see https://docs.ansible.com/ansible-core/2.15/playbook_guide/playbooks_privilege_escalation.html#risks-of-becoming-an-unprivileged-user#risks-of-becoming-an-unprivileged-user"}
```

I don't know what to do now..

**Comment by richardlau** [comment]
I'm seeing the same error too 😞

**Comment by richardlau** [comment]
Wonder if it's related to https://github.com/ansible/ansible/issues/79602

**Comment by targos** [comment]
I confirm that installing `acl` fixed the issue.

**Comment by targos** [comment]
Docker images are now building.

`test-osuosl-centos7_container-arm64-1` already failed with:

```
Collecting tap2junit==0.1.6", "\u001b[91m  Could not find a version that satisfies the requirement tap2junit==0.1.6 (from versions: 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.1.0, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.5)"
```

**Comment by richardlau** [comment]
> Docker images are now building.
> 
> `test-osuosl-centos7_container-arm64-1` already failed with:
> 
> ```
> Collecting tap2junit==0.1.6", "\u001b[91m  Could not find a version that satisfies the requirement tap2junit==0.1.6 (from versions: 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.1.0, 0.1.1, 0.1.2, 0.1.3, 0.1.4, 0.1.5)"
> ```

That is likely related to the version of Python 3 being used. tap2junit 0.1.6 requires Python 3.7: https://pypi.org/project/tap2junit/. I had to use a non-system Python 3 for RHEL 8, for example, https://github.com/nodejs/build/issues/3306.

**Review by richardlau — APPROVED** [review]

**Comment by targos** [comment]
There's another broken docker build with `test-osuosl-ubuntu1804_sharedlibs_container-arm64-1`:

```
"Step 30/36 : ENV ZLIBVER 1.2.13", " ---> Using cache", " ---> dae1cdad655e", "Step 31/36 : ENV ZLIB12DIR /opt/zlib_$ZLIBVER", " ---> Using cache", " ---> b11991c97718", "Step 32/36 : RUN mkdir -p /tmp/zlib_$ZLIBVER &&     cd /tmp/zlib_$ZLIBVER &&     curl -sL https://zlib.net/zlib-$ZLIBVER.tar.gz | tar zxv --strip=1 &&     ./configure --prefix=$ZLIB12DIR &&     make -j 6 &&     make install &&     rm -rf /tmp/zlib_$ZLIBVER", " ---> Running in 55c89873c059", "\u001b[91m", "gzip: stdin: not in gzip format", "\u001b[0m\u001b[91mtar: Child returned status 1", "tar: Error is not recoverable: exiting now", "\u001b[0m"]}
```

I'm not sure how to proceed at this point, as the playbook stops because of those two build failures.

**Comment by richardlau** [comment]
> There's another broken docker build with `test-osuosl-ubuntu1804_sharedlibs_container-arm64-1`:
> 
> ```
> "Step 30/36 : ENV ZLIBVER 1.2.13", " ---> Using cache", " ---> dae1cdad655e", "Step 31/36 : ENV ZLIB12DIR /opt/zlib_$ZLIBVER", " ---> Using cache", " ---> b11991c97718", "Step 32/36 : RUN mkdir -p /tmp/zlib_$ZLIBVER &&     cd /tmp/zlib_$ZLIBVER &&     curl -sL https://zlib.net/zlib-$ZLIBVER.tar.gz | tar zxv --strip=1 &&     ./configure --prefix=$ZLIB12DIR &&     make -j 6 &&     make install &&     rm -rf /tmp/zlib_$ZLIBVER", " ---> Running in 55c89873c059", "\u001b[91m", "gzip: stdin: not in gzip format", "\u001b[0m\u001b[91mtar: Child returned status 1", "tar: Error is not recoverable: exiting now", "\u001b[0m"]}
> ```
> 
> I'm not sure how to proceed at this point, as the playbook stops because of those two build failures.

Ah, this one is https://github.com/nodejs/build/pull/3067#issuecomment-1303772931 -- zlib only keep the most recent version available to download from https://zlib.net/ but older ones are available on https://www.zlib.net/fossils/. Apparently I updated the docker templates for the Ubuntu 20.04 sharedlibs containers in https://github.com/nodejs/build/pull/3076/files#diff-12b3498626d53cb9267e5ec7cec164c7bf5a3d703c0239277071edac9f3b69da but not the Ubuntu 18.04 based ones.

@targos if you like I can take a look at fixing the two failing containers today after lunch.

**Comment by targos** [comment]
> if you like I can take a look at fixing the two failing containers today after lunch.

That would be awesome. I won't have much more time for this today

**Comment by richardlau** [comment]
I've pushed two commits to this PR to fix the issues with the CentOS 7 container and Ubuntu 18.04 sharedlibs container.
@targos FYI we might also need to add ICU 73 to the sharedlib containers (either as part of this PR or a follow-up) in preparation for https://github.com/nodejs/node/pull/49639.

**Review by targos — COMMENTED** [review]
I can't approve because it's my PR, but @richardlau's commits LGTM. Thanks for the help!

**Review by MoLow — APPROVED** [review]

**Comment by targos** [comment]
It looks like something's wrong: https://ci.nodejs.org/job/node-test-commit-arm-debug/9398/nodes=ubuntu2004_debug-arm64/console

**Comment by targos** [comment]
```
18:08:19 + curl -sLO https://raw.githubusercontent.com/nodejs/build/main/jenkins/scripts/select-compiler.sh
18:08:19 + . ./select-compiler.sh
18:08:19 ++ '[' '' '!=' DONT ']'
18:08:19 ++ NODE_NAME=test-equinix-ubuntu2004_sharedlibs_container-arm64-5
18:08:19 ++ echo 'Selecting compiler based on test-equinix-ubuntu2004_sharedlibs_container-arm64-5'
18:08:19 Selecting compiler based on test-equinix-ubuntu2004_sharedlibs_container-arm64-5
18:08:19 ++ case $NODE_NAME in
18:08:19 ++ SELECT_ARCH=ARM64
18:08:19 ++ '[' -z x ']'
18:08:19 ++ case $NODE_NAME in
18:08:19 ++ '[' ARM64 = PPC64LE ']'
18:08:19 ++ '[' ARM64 = S390X ']'
18:08:19 ++ '[' ARM64 = IBMI73 ']'
18:08:19 ++ '[' ARM64 = AIXPPC ']'
18:08:19 ++ '[' ARM64 = X64 ']'
18:08:19 ++ '[' ARM64 = ARM64 ']'
18:08:19 ++ echo 'Setting compiler for Node version 21 on arm64'
18:08:19 Setting compiler for Node version 21 on arm64
18:08:19 ++ case $nodes in
18:08:19 + export 'CONFIG_FLAGS= --debug'
18:08:19 + CONFIG_FLAGS=' --debug'
18:08:19 + NODE_TEST_DIR=/home/iojs/node-tmp
18:08:19 + FLAKY_TESTS=dontcare
18:08:19 + make run-ci -j 16
18:08:19 python3 ./configure --verbose  --debug
18:08:21 Node.js configure: Found Python 3.8.10...
18:08:21 Detected C++ compiler (CXX=ccache g++) version: 9.4.0
18:08:21 WARNING: C++ compiler (CXX=ccache g++, 9.4.0) too old, need g++ 10.1.0 or clang++ 8.0.0
18:08:21 Detected C compiler (CC=ccache gcc) version: 9.4.0
```

**Comment by richardlau** [comment]
@targos ah from https://ci.nodejs.org/job/node-test-commit-arm-debug/9398/nodes=ubuntu2004_debug-arm64/injectedEnvVars/ the value of `nodes` is `ubuntu2004_debug-arm64` which isn't matching https://github.com/nodejs/build/blob/12ecb2d2c6ee57d31f128c35f93108a58b2a7d58/jenkins/scripts/select-compiler.sh#L242
I guess we've actually run into https://github.com/nodejs/build/pull/3485#discussion_r1324716173 🙂.

**Comment by richardlau** [comment]
Update to the compiler selector script: https://github.com/nodejs/build/pull/3493

**Pull request merged**